File: ascenda.py

#
# Complete the 'merchant_offers' function below.
#
# The function is expected to return a STRING.
# The function accepts following parameters:
#  1. STRING checkin
#  2. STRING offers_json
#
import json
import copy
from datetime import datetime, timedelta

def remove_invalid_offers(checkin, offers):
    '''
    Function to filter offers based on category and their required rules
    Restaurant: The offer is valid until the check-in date plus 3 days.
    Retail: The offer is valid until the check-in date plus 5 days.
    Activity: The offer is valid until the check-in date plus 7 days.
    '''
    checkin_dt = datetime.strptime(checkin, "%Y-%m-%d")
    filtered_offers = []

    for offer in offers:
        valid_to_dt = datetime.strptime(offer["valid_to"], "%Y-%m-%d")
        match offer['category']:
            case 1:
                # Resturant
                if checkin_dt + timedelta(days=3) <= valid_to_dt:
                    filtered_offers.append(offer)
            case 2: 
                # Retail
                if checkin_dt + timedelta(days=5) <= valid_to_dt:
                    filtered_offers.append(offer)
            case 4:
                # Activity
                if checkin_dt + timedelta(days=7) <= valid_to_dt:
                    filtered_offers.append(offer)
            case _:
                continue

    return filtered_offers

def get_min_merchant(merchant):
    return merchant['distance']

def get_min_distance_for_each_cat(offers):
    '''
    Function to sort offers based on their distance, incrementally
    '''
    copied_offers = offers

    for offer in copied_offers:
      offer["merchants"] = min(offer["merchants"], key = get_min_merchant)

    sorted_offers = sorted(copied_offers, key = lambda offer: (offer["merchants"]["distance"]))
    return sorted_offers

def get_sorted_offers_with_score(offers, age_group, gender):
    '''
    Function to sort offers by highest scores
    '''
    age_score = 1.25
    gender_score = 1.5
    copied_offers = offers

    for offer in copied_offers:
        offer["gender_scores"] = offer["gender_scores"][gender] * gender_score
        offer["age_scores"] = offer["age_scores"][age_group] * age_score
    
    sorted_offers = sorted(copied_offers, key = lambda offer: (offer["gender_scores"] + offer["age_scores"]), reverse=True)
    return sorted_offers

def format_offer(offer):
    match offer["category"]:
        case 1:
            return {
                "id": offer["id"],
                "title": offer["title"],
                "description": offer["description"],
                "category": "Restaurant"
            }
        case 2:
            return {
                "id": offer["id"],
                "title": offer["title"],
                "category": "Retail"
            }
        case 4:
            return {
                "id": offer["id"],
                "title": offer["title"],
                "description": offer["description"],
                "valid_to": offer["valid_to"],
                "category": "Activity",
            }
    

def get_best_offers(offers):
    '''
    Function to get 2 offers of different categories
    '''
    have_category = None
    ret = []

    for offer in offers:
        filtered_offer = format_offer(offer)
        if have_category == None:
            # no offers taken
            ret.append(filtered_offer)

            have_category = offer['category']
            continue
        if have_category == offer['category']:
            continue
        ret.append(filtered_offer)
        break
    return ret        

def merchant_offers(checkin, offers_json, age_group, gender):
    '''
    Function to filter and get best offers
    '''
        
    valid_offers = remove_invalid_offers(checkin, offers_json)
    offers_with_score = get_sorted_offers_with_score(valid_offers, age_group, gender)
    cat_offers_min_distance = get_min_distance_for_each_cat(offers_with_score)
    best_offers = get_best_offers(cat_offers_min_distance)

    return best_offers

if __name__ == '__main__':
    
  with open('offers.json', 'r') as f:
    data = json.load(f)



  # print(merchant_offers('2023-05-22', copy.deepcopy(data)))
  # print(merchant_offers('2023-05-20', copy.deepcopy(data)))
  # print(merchant_offers('2023-05-18', copy.deepcopy(data)))
  # print(merchant_offers('2023-05-15', copy.deepcopy(data)))

  # print(merchant_offers('2023-05-15', copy.deepcopy(data), 'young_adults', 'male'))   # [3, 6]
  # print(merchant_offers('2023-05-15', copy.deepcopy(data), 'teens', 'male'))          # [3, 2]
  # print(merchant_offers('2023-05-15', copy.deepcopy(data), 'seniors', 'male'))        # [2, 3]
    
  print(merchant_offers('2023-05-15', copy.deepcopy(data), 'young_adults', 'male'))     # [{'id': 3, 'title': 'Offer 3', 'description': 'Offer 3 description', 'valid_to': '2023-05-24', 'category': 'Activity'}, {'id': 6, 'title': 'Offer 6', 'category': 'Retail'}]

File: ascenda.ts
import * as fs from "fs"

// Type definitions
export interface Offer {
  id: number
  title: string
  description: string
  category: number
  merchants: Array<Merchant>
  age_scores: AgeScore
  gender_scores: GenderScore
  valid_to: string
  total_score?: number
  merchant_min_dist?: Merchant
}

export interface Merchant {
  id: number
  name: string
  distance: number
}

export interface AgeScore {
  adults: number
  seniors: number
  young_adults: number
  teens: number
}

export interface GenderScore {
  male: number
  female: number
  non_binary: number
}

export interface OfferFormat {
  id: number
  title: string
  category: string
  description?: string
  valid_to?: string
}

type ageGroup = 'adults' | 'seniors' | 'young_adults' | 'teens'
type genderGroup = 'male' | 'female' | 'non_binary'

// Main class
export class OfferService {
  merchantOffers(checkin: string, offers: Array<Offer>, ageGroup: ageGroup, gender: genderGroup): Array<OfferFormat> {
    const filteredOffers = offers.filter((offer) => this.filterInvalidOffers(offer, checkin))

    filteredOffers.forEach((offer) => {
      const totalScore = offer.age_scores[ageGroup] * 1.25  + offer.gender_scores[gender] * 1.5
      offer.total_score = totalScore
      offer.merchant_min_dist = offer.merchants.reduce((min, current) => (current.distance < min.distance ? current : min), offer.merchants[0])
    })

    filteredOffers.sort(this.sortOffersByDistAndScore)

    const bestOffers = this.get2BestOffer(filteredOffers)

    return bestOffers
  }

  private filterInvalidOffers(offer: Offer, checkin: string): boolean {
    const checkinDate = new Date(checkin).getTime()
    const offerValidToDate = new Date(offer.valid_to).getTime()
    const oneDay = 24 * 60 * 60 * 1000; // hr * min * sec * ms
    const dayDifference = Math.floor((offerValidToDate - checkinDate) / oneDay)
  
    switch (offer.category) {
      case 1: {
        // Restaurant
        if (dayDifference > 3) {
          return true
        }
        break
      }
      case 2: {
        // Retail
        if (dayDifference > 5) {
          return true
        }
        break
      }
      case 4: {
        // Activity
        if (dayDifference > 7) {
          return true
        }
        break
      }
      default: {
        break
      }
    }
    return false
  }

  private sortOffersByDistAndScore(offer1: Offer, offer2: Offer): number {
    // Lower Distance priority, followed by highest score
    if (offer1.merchant_min_dist!.distance !== offer2.merchant_min_dist!.distance) {
      return offer1.merchant_min_dist!.distance - offer2.merchant_min_dist!.distance
    }
    return offer2.total_score! - offer1.total_score!
  }

  private get2BestOffer(offers: Array<Offer>): Array<OfferFormat> {
    let currentCategory = -1 // Since we only have 4, I assume -1 is assigned as none
    const bestOffers: Array<OfferFormat> = []

    for (var offer of offers) {
      const formattedOffer = this.formatOffer(offer)

      if (currentCategory == -1) {
        bestOffers.push(formattedOffer)
        currentCategory = offer.category
        continue
      } 
      if (currentCategory != offer.category) {
        bestOffers.push(formattedOffer)
        break
      }
    }
    return bestOffers
  }

  private formatOffer(offer: Offer): OfferFormat {
    switch (offer.category) {
      case 1: {
        return {
          id: offer.id,
          title: offer.title,
          description: offer.description,
          category: "Restaurant"
        }
      }
      case 2: {
        return {
          id: offer.id,
          title: offer.title,
          category: "Retail"
        }
      }
      case 4: {
        return {
          id: offer.id,
          title: offer.title,
          description: offer.description,
          valid_to: offer.valid_to,
          category: "Activity"
        }
      }
      default: {
        break
      }
    }
    throw new Error('Unexpected category type, expected 1, 2 or 4 only')
  }

}

const rawData = fs.readFileSync("offers.json", "utf8")
const jsonData: Array<Offer> = JSON.parse(rawData)
const offerService = new OfferService()

// offerService.merchantOffers('2023-05-15', jsonData, 'young_adults', 'male')
// offerService.merchantOffers('2023-05-15', jsonData, 'teens', 'male')
// offerService.merchantOffers('2023-05-15', jsonData, 'seniors', 'female')
console.log(offerService.merchantOffers('2023-05-15', jsonData, 'young_adults', 'male'))
